Automattic/woocommerce-payments
# Prevent webhook duplicate actions

Fixes #3814 
See more details in this Rado's comment https://github.com/Automattic/woocommerce-payments/issues/3814#issuecomment-1066691553

#### Changes proposed in this Pull Request

I did research for all webhook event types below. 

To resolve for most potential duplicate actions, I created a new method: `WC_Payments_Order_Service::order_note_exists`, and only allowed processing data if the note does not exist. 

Note: 

- I did not only check for a unique string as it's not reliable. For example, if I use the dispute ID to check if it exists in the order notes, multiple dispute webhooks with different notes can not be added. 









#### Research 

| Event Type | Before  | After |
| ------- | ------- | --------|
| charge.refund.updated | Focus on `failed` charge only, and no check to avoid duplicate actions | ~[Asking here](https://github.com/Automattic/woocommerce-payments/pull/3965/files#r827772755), handled by @achyuthajoy~ Fixed with `order_note_exists` |
| charge.dispute.created |  Add a note and update status, but no check to avoid duplicate actions  |  Fixed with `order_note_exists` |
| charge.dispute.closed |  Add a note and update status, but no check to avoid duplicate actions  |  Fixed with `order_note_exists` |
| charge.dispute.funds_reinstated |  Only add a note, but no check to avoid duplicate actions  |  Fixed with `order_note_exists` |
| charge.dispute.funds_withdrawn |  Only add a note, but no check to avoid duplicate actions  |  Fixed with `order_note_exists` |
| charge.dispute.updated |  Only add a note, but no check to avoid duplicate actions  |  Fixed with `order_note_exists` |
| charge.expired |  Add a note and update status, but no check to avoid duplicate actions  |  Fixed with `order_note_exists` |
| account.updated | It's OK to have duplicate actions, i.e. refresh account data  |  No change |
| wcpay.notification | Current checks have already prevented duplicate actions. See further in `put_note` and `can_note_be_added` of WC_Payments_Remote_Note_Service  |  No change |
| payment_intent.payment_failed |  Seems not enough as it applies for two payment types: `au_becs_debit` and `us_bank_account`, and checks the existing order status. |  Fixed with `order_note_exists` |
| payment_intent.succeeded |  Seems not enough and it checks the order status before processing only |  Fixed with `order_note_exists` |
| invoice.upcoming |  It seems not under my eyes but needs the check from Helix |  Ask Helix |
| invoice.paid |  It seems not under my eyes but needs the check from Helix |  Ask Helix |
| invoice.payment_failed |  It seems not under my eyes but needs the check from Helix |  Ask Helix |


#### Testing instructions





* Turn on your webhook listen for your local server.
* Use the dispute test cards as mentioned [here](https://stripe.com/docs/testing#disputes). 
* See the expected note is created in the order management (in WP Admin).
* Try to resend webhooks for your server.
* Confirm: no new note is created for the same webhook.

-------------------


**Post merge**



- [ ] Create an issue for Helix to handle subscription events.
